### PR TITLE
Update hypre package

### PR DESCRIPTION
**Pull request by victorapm**
1. Add MAGMA support to hypre (available since hypre v2.29.0)
2. Allow compilation with cuda >= 12 (available since hypre v2.29.0)
3. Minor fix related to the use of `cache_extra_test_sources`

### CODE REVIEW

File: var/spack/repos/builtin/packages/hypre/package.py [hunk 1]
```diff
@@ -329,7 +344,7 @@ def install(self, spec, prefix):
 
     @run_after("install")
     def cache_test_sources(self):
-        self.cache_extra_test_sources(self.extra_install_tests)
+        cache_extra_test_sources(self, self.extra_install_tests)
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/hypre/package.py:347]: ~Why was this change needed?~ Don't mind this comment, I just saw the function got deprecated.

**Review by alalazo — APPROVED** [review]

**Comment by alalazo** [comment]
Giving a couple of days more for maintainers to comment on the PR. Feel free to ping me for a merge at any time.

**Review by osborn9 — APPROVED** [review]
Thanks @victorapm. Looks good!
